File: regions/system/library/go/distributed-lock/redis_lock.go
package distributedlock

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"time"

	"github.com/redis/go-redis/v9"
)

// releaseScript is a Lua script for safe lock release.
// Only deletes the key if the stored value matches the token.
const releaseScript = `
if redis.call("get", KEYS[1]) == ARGV[1] then
    return redis.call("del", KEYS[1])
else
    return 0
end
`

// extendScript is a Lua script for safe lock extension.
// Only extends TTL if the stored value matches the token.
const extendScript = `
if redis.call("get", KEYS[1]) == ARGV[1] then
    return redis.call("pexpire", KEYS[1], ARGV[2])
else
    return 0
end
`

// RedisLock は Redis を使用した分散ロック実装。
// SET NX PX によるアトミックなロック取得と、Lua スクリプトによる安全な解放を実現する。
type RedisLock struct {
	client    redis.Cmdable
	keyPrefix string
}

// RedisLockOption は RedisLock の設定オプション。
type RedisLockOption func(*RedisLock)

// WithLockPrefix はロックキーのプレフィックスを設定する。
func WithLockPrefix(prefix string) RedisLockOption {
	return func(l *RedisLock) {
		l.keyPrefix = prefix
	}
}

// NewRedisLock は新しい RedisLock を生成する。
func NewRedisLock(client redis.Cmdable, opts ...RedisLockOption) *RedisLock {
	l := &RedisLock{
		client:    client,
		keyPrefix: "lock",
	}
	for _, opt := range opts {
		opt(l)
	}
	return l
}

// NewRedisLockFromURL は Redis URL から新しい RedisLock を生成する。
func NewRedisLockFromURL(url string, opts ...RedisLockOption) (*RedisLock, error) {
	options, err := redis.ParseURL(url)
	if err != nil {
		return nil, err
	}
	client := redis.NewClient(options)
	return NewRedisLock(client, opts...), nil
}

func (l *RedisLock) lockKey(key string) string {
	return l.keyPrefix + ":" + key
}

func (l *RedisLock) Acquire(ctx context.Context, key string, ttl time.Duration) (*LockGuard, error) {
	fullKey := l.lockKey(key)
	token := generateRedisToken()

	ok, err := l.client.SetNX(ctx, fullKey, token, ttl).Result()
	if err != nil {
		return nil, err
	}
	if !ok {
		return nil, ErrAlreadyLocked
	}

	return &LockGuard{Key: key, Token: token}, nil
}

func (l *RedisLock) Release(ctx context.Context, guard *LockGuard) error {
	fullKey := l.lockKey(guard.Key)

	result, err := l.client.Eval(ctx, releaseScript, []string{fullKey}, guard.Token).Int64()
	if err != nil {
		return err
	}
	if result == 0 {
		return ErrTokenMismatch
	}
	return nil
}

// Extend はロックの有効期限を延長する。ガードのトークンが一致する場合のみ延長する。
func (l *RedisLock) Extend(ctx context.Context, guard *LockGuard, ttl time.Duration) error {
	fullKey := l.lockKey(guard.Key)
	millis := ttl.Milliseconds()

	result, err := l.client.Eval(ctx, extendScript, []string{fullKey}, guard.Token, millis).Int64()
	if err != nil {
		return err
	}
	if result == 0 {
		return ErrTokenMismatch
	}
	return nil
}

func (l *RedisLock) IsLocked(ctx context.Context, key string) (bool, error) {
	fullKey := l.lockKey(key)
	count, err := l.client.Exists(ctx, fullKey).Result()
	if err != nil {
		return false, err
	}
	return count > 0, nil
}

func generateRedisToken() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

File: regions/system/library/typescript/distributed-lock/src/redis-lock.ts
import { randomBytes } from 'crypto';
import type { Redis } from 'ioredis';
import { LockError, type LockGuard } from './index.js';

const RELEASE_SCRIPT = `
if redis.call("get", KEYS[1]) == ARGV[1] then
  return redis.call("del", KEYS[1])
else
  return 0
end
`;

const EXTEND_SCRIPT = `
if redis.call("get", KEYS[1]) == ARGV[1] then
  return redis.call("pexpire", KEYS[1], ARGV[2])
else
  return 0
end
`;

export class RedisDistributedLock {
  private readonly keyPrefix: string;

  constructor(
    private readonly redis: Redis,
    keyPrefix = 'lock',
  ) {
    this.keyPrefix = keyPrefix;
  }

  private lockKey(key: string): string {
    return `${this.keyPrefix}:${key}`;
  }

  async acquire(key: string, ttlMs: number): Promise<LockGuard> {
    const fullKey = this.lockKey(key);
    const token = randomBytes(16).toString('hex');

    const result = await this.redis.set(fullKey, token, 'PX', ttlMs, 'NX');

    if (result === null) {
      throw new LockError(`lock already held: ${key}`, 'LOCK_HELD');
    }

    return { key, token };
  }

  async release(guard: LockGuard): Promise<void> {
    const fullKey = this.lockKey(guard.key);
    const result = await this.redis.eval(RELEASE_SCRIPT, 1, fullKey, guard.token) as number;

    if (result !== 1) {
      throw new LockError(`lock not owned: ${guard.key}`, 'NOT_OWNER');
    }
  }

  async extend(guard: LockGuard, ttlMs: number): Promise<void> {
    const fullKey = this.lockKey(guard.key);
    const result = await this.redis.eval(EXTEND_SCRIPT, 1, fullKey, guard.token, String(ttlMs)) as number;

    if (result !== 1) {
      throw new LockError(`lock not owned: ${guard.key}`, 'NOT_OWNER');
    }
  }

  async isLocked(key: string): Promise<boolean> {
    const fullKey = this.lockKey(key);
    const exists = await this.redis.exists(fullKey);
    return exists === 1;
  }
}
